src/core/dappletContext.ts

```diff
@@ -20,20 +21,70 @@ export class DappletContext {
 
     constructor(config: ContextConfig = DEFAULT_CONFIG) {
         config = { ...DEFAULT_CONFIG, ...config };
-        this.dappletProviders = config.providers!;
+        this.dappletProviders = config.providers! || [];
         this.featureRegistry = new FeatureRegistry(config.features || []);
     }
 
     async processRequest(request: DappletRequest): Promise<DappletTxResult> {
+        // dapplet loading and validation
+        for (let i = 0; i < request.frames.length; i++) {
+            const dapplet = await this._loadDapplet(request.frames[i].dappletId);
+            this._validateDapplet(dapplet);
+            request.frames[i].dapplet = dapplet;
+        }
+
         const engine = new DappletEngine(request, this);
-        engine.onStatusChanged((statuses) => console.log(statuses.map(s => FrameStatus[s]).join(", ")));
-        await engine.load();
-        engine.validate();
-        
+        engine.run();
+
         //const activity = new DappletActivity(request, this);
         return {};
```

Comment:
we probably need a handle as a return value to subscribe for events from this request.
Let us discuss it,